src/cascadia/TerminalSettingsModel/IInheritable.h

```diff
@@ -0,0 +1,215 @@
+/*++
+Copyright (c) Microsoft Corporation
+Licensed under the MIT license.
+
+Module Name:
+- IInheritable.h
+
+Abstract:
+- An interface allowing settings objects to inherit settings from a parent
+
+Author(s):
+- Carlos Zamora - October 2020
+
+--*/
+#pragma once
+
+namespace winrt::Microsoft::Terminal::Settings::Model::implementation
+{
+    template<typename T>
+    struct IInheritable
+    {
+    public:
+        // Method Description:
+        // - Create a new instance of T, but set its parent to this instance
+        // Arguments:
+        // - <none>
+        // Return Value:
+        // - a new instance of T with this instance set as its parent
+        com_ptr<T> CreateChild() const
+        {
+            auto child{ winrt::make_self<T>() };
+
+            com_ptr<T> parent;
+            winrt::copy_from_abi(parent, const_cast<T*>(static_cast<const T*>(this)));
+            child->InsertParent(parent);
+
+            child->_FinalizeInheritance();
+            return child;
+        }
+
+        void InsertParent(com_ptr<T> parent)
+        {
+            _parents.push_back(parent);
+        }
+
+        void InsertParent(size_t index, com_ptr<T> parent)
+        {
+            auto pos{ _parents.begin() };
+            for (size_t i=0; i<index; ++i)
+            {
+                ++pos;
+            }
+            _parents.insert(pos, parent);
+        }
+
+    protected:
+        std::vector<com_ptr<T>> _parents{};
+
+        // Method Description:
+        // - Actions to be performed after a child was created. Generally used to set
+        //   any extraneous data from the parent into the child.
+        // Arguments:
+        // - <none>
+        // Return Value:
+        // - <none>
+        virtual void _FinalizeInheritance() {}
+    };
+
+    // This is like std::optional, but we can use it in inheritance to determine whether the user explicitly cleared it
+    template<typename T>
+    struct NullableSetting
+    {
+        winrt::Windows::Foundation::IReference<T> setting{ nullptr };

```

Comment:
if you use `std::optional` even inside NullableSetting -- for the same exact reason as you'd use it outside -- you don't need a special string type. Remember: IReference can be our _api contract_ but it doesn't need to pervade our internal implementations

Reply:
Problem is, then we'd have to convert between `nullopt` and `nullptr`. I'd rather just keep it saved as an `IReference` internally. That way, when we get/set the value (and it's valid) we just return `NullableSetting.setting`, instead of translating it on the way in/out.